rfcs/text/0014_api_documentation.md

```diff
@@ -0,0 +1,356 @@
+are huge, and can't be rendered in a single page.
+
+
+```ts
+export interface PluginApi {
+  id: string;
+  serviceFolders?: readonly string[];
+  client: ScopeApi;
+  server: ScopeApi;
+  common: ScopeApi;
+}
+```
+
+## Manifest file changes
+
+**serviceFolders?: string[]**
+
+Used by the system to group services into sub-pages. Some plugins, like data and core, have such huge APIs they are very slow to contain in a single page, and they are less consummable by solution developers. The addition of an optional list of services folders will cause the system to automatically create a separate page with every API that is defined within that folder. The caveat is that core will need to define a manifest file in order to define its service folders...
+
+**Using a kibana.json file for core**
+
+For the purpose of API infrastructure, core is treated like any other plugin. This means it has to specify serviceFolders section inside a manifest file to be split into sub folders. There are other ways to tackle this - like a hard coded array just for the core folder, but I kept the logic as similar to the other plugins as possible.
+
+**teamOwner: string**
+
+Team owner can be determined via github CODEOWNERS file, but we want to encourage single team ownership per plugin. Requiring a team owner string in the manifest file will help with this and will allow the API doc system to manually add a section to every page that has a link to the team owner. Additional ideas are teamSlackChannel or teamEmail for further contact. 
+
+**summary: string**
+
+A brief description of the plugin can then be displayed in the automatically generated API documentation.
+
+## Technology: ts-morph vs api-extractor
+
+[Api-extractor](https://api-extractor.com/) is a utility built from microsoft that parses typescript code into json files that can then be used in a custom [api-documenter](https://api-extractor.com/pages/setup/generating_docs/) in order to build documentation. This is what we [have now](https://github.com/elastic/kibana/tree/master/docs/development), except we use the default api-documenter. Unfortunately, because our plugins aren’t separate packages, and the way api-extractor works, we can’t build cross plugin links this way.
+
+It is unknown whether the work with bazel will allow us to use this technology. I suspect not, because even though there will be separate `.d.ts` files, the plugins will still be importing types via relative links.
+
+[ts-morph](https://github.com/dsherret/ts-morph) is a utility built and maintained by a single person, which sits a layer above the raw typescript compiler. It affords greater flexibility, thus supports cross plugin links (among other things like links to source files). The downsides of using this library are:
+
+ - Risks of relying on a package maintained by a single developer
+ - Less re-usability across repositories. What if EUI wanted to use the same system?  
```

Comment:
I'll note your concern in the RFC, I agree it's an important one.

The author does seem to regularly keep up with typescript upgrades, but there is still risk that he will slow down, or certain upgrades will require a large effort on our part.

> In my opinion more expensive than not having docs, but this is hard to quantify

We always have the option of making this call in the future. Each time we prepare for a typescript upgrade, we'll need to decide between:

1. Support the new version in our ts-morph implementation 
2. Switch to an api-extractor implementation
3. Remove our API docs

If it's easy to upgrade ts-morph, we make the changes and move on. If it's not, we assess how useful we think the docs are compared to the effort it would take to keep them. If the effort outweighs their usefulness, we can remove them, perhaps temporarily, until there is more bandwidth to fix the issues and bring back support.

If you're worried about devs starting to rely on docs that we might need to stop supporting in order to move faster with typescript, I think that would indicate a high level of importance, and mean it's worth investing the effort into keeping them around.